### DISCUSSION TITLE
Possible to do online/streaming inference with neural mvdr beamformer?

### DISCUSSION THREAD

**Discussion by Jacob-Bishop** [espnet/espnet#5207] [Q&A] [answered]
I'm looking at the pretrained neural mvdr beamformer example in https://colab.research.google.com/drive/1fjRJCh96SoYLZPRxsjF9VDv4Q2VoIckI (espnet/Wangyou_Zhang_chime4_enh_train_enh_beamformer_mvdr_raw).

Is it possible to use this model for online processing or something close to it? I've tried swapping out SeparateSpeech for SeparateSpeechStreaming, but got a not implemented error. I also tried chunking the audio, which technically worked, but even 0.1 second chunks had bad performance.
[Edit: I also tried adding segment_size=2.4, hop_size=0.8, and normalize_segment_scale=False options for SeparateSpeech. It does appear to work with 1 second chunks.]

If this isn't possible, is there a recommended online beamformer in espnet?

**Comment by sw005320** [comment]
@Emrys365, can you add some comments about this?
I think the shorter chunk size is one option.
We can also think of supporting pure online processing in the future.

**Comment by Emrys365** [comment, marked as answer]
> Is it possible to use this model for online processing or something close to it?

Unfortunately, it is not supported now.
Online processing requires dynamically calculating and updating the beamformer filter in different frames.
It is not straightforward to use the current implementation to achieve online processing, since the current one calculates the beamformer filter using all available snapshots in the full-length input and it remains unchanged across frames.

> I also tried chunking the audio, which technically worked, but even 0.1 second chunks had bad performance.

I would expect it to work if you retrain or train the model on a similar length setting, i.e. using the same chunk length for training. The provided model was trained on full-length samples, and it consists of BLSTMs in the architecture, so it might be better to retrain it if you want to use the chunking method for block-online processing.

**Reply by Jacob-Bishop** [reply]
Thanks for the detailed writeup @Emrys365. Good to know I'm not just holding it wrong...

**Comment by Emrys365** [comment]
> If this isn't possible, is there a recommended online beamformer in espnet?

As of now, we haven't implemented online versions of the beamformers in ESPnet. I will add it to the development plan (https://github.com/espnet/espnet/issues/2200).
That being said, you could still refer to existing implementations somewhere else, such as https://github.com/funcwj/setk/blob/master/scripts/sptk/libs/beamformer.py#L286.
